# 4.5 Exercise

In [1]:
import numpy as np
import pandas as pd
import os

In [2]:
# file path
path = r'/Users/elliothaigh/05-2023 Instacart Basket Analysis'

In [3]:
# import data sets
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', '4.3_orders_products', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# create a dataframe

df_test = pd.DataFrame()

In [5]:
df_test['mix'] = ['a', 'b', '1', True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# test for mixed columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Search for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# removing missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# creating subset for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
# dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [17]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_prods_clean_no_dups.csv'))

# Task 4.5

In [18]:
# describe statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

Statistics for df_orders appear to be as expected

In [35]:
# Check for mixed columns in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, 'mixed')
  else: print (col, 'not mixed')

Unnamed: 0 not mixed
order_id not mixed
user_id not mixed
order_number not mixed
order_day_of_week not mixed
order_hour_of_day not mixed
days_since_prior_order not mixed


In [36]:
# Search for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are only missing values (206209) in the days since prior order column. This could be becasue these orders were the first by the customers and thus there were no prior orders to count. 

In [37]:
# Create subset for missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [38]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [49]:
# add first order column to indicate if the order is the first for the customer
df_ords['first_order'] = ['No' if x >= 0 else 'Yes' for x in df_ords.days_since_prior_order]

We have left the null values as they are as they indicate that there have been no prior orders. Instead we added a new column to indicate where or not an order was the first for a customer.

In [50]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
1                 1   2398795        1             2                  3   
2                 2    473747        1             3                  3   
3                 3   2254736        1             4                  4   
4                 4    431534        1             5                  4   
...             ...       ...      ...           ...                ...   
3421078     3421078   2266710   206209            10                  5   
3421079     3421079   1854736   206209            11                  4   
3421080     3421080    626363   206209            12                  1   
3421081     3421081   2977660   206209            13                  1   
3421082     3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_prior_order first_order  
0                        8                     NaN         Yes  
1                        7                    15.0          No  
2                       12                    21.0          No  
3                        7                    29.0          No  
4                       15                    28.0          No  
...                    ...                     ...         ...  
3421078                 18                    29.0          No  
3421079                 10                    30.0          No  
3421080                 12                    18.0          No  
3421081                 12                     7.0          No  
3421082                 14                    30.0          No  

[3421083 rows x 8 columns]

In [51]:
# Checking for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [52]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

No duplicates found

In [53]:
# Export df_ords
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'df_ords_clean_no_dups.csv'))